x-pack/legacy/plugins/actions/server/builtin_action_types/webhook.ts

```diff
@@ -0,0 +1,154 @@
+  path: nullableType(schema.string()),
+  scheme: schema.oneOf(
+    [schema.literal(WebhookSchemes.HTTP), schema.literal(WebhookSchemes.HTTPS)],
+    {
+      defaultValue: WebhookSchemes.HTTP,
+    }
+  ),
+});
+
+const ConfigSchema = schema.object({
+  url: schema.oneOf([schema.string(), CompositeUrlSchema]),
+  method: schema.oneOf([schema.literal(WebhookMethods.POST), schema.literal(WebhookMethods.PUT)], {
+    defaultValue: WebhookMethods.POST,
+  }),
+  headers: nullableType(HeadersSchema),
+});
+
+// secrets definition
+export type ActionTypeSecretsType = TypeOf<typeof SecretsSchema>;
+const SecretsSchema = schema.object({
+  user: schema.string(),
+  password: schema.string(),
+});
+
+// params definition
+export type ActionParamsType = TypeOf<typeof ParamsSchema>;
+const ParamsSchema = schema.object({
+  body: schema.maybe(schema.string()),
+});
+
+// action type definition
+export const actionType: ActionType = {
+  id: '.webhook',
+  name: 'webhook',
+  validate: {
+    config: ConfigSchema,
+    secrets: SecretsSchema,
+    params: ParamsSchema,
+  },
+  executor,
+};
+
+function asComposedUrl(url: string | TypeOf<typeof CompositeUrlSchema>): string {
+  if (typeof url === 'string') {
+    return url;
+  }
+  return `${url.scheme}://${url.host}${url.port ? `:${url.port}` : ''}${url.path || ''}`;
+}
+
+function describeWebhookErrorSource(error: AxiosError): string {
+  if (error.response) {
+    // The request was made and the server responded with a status code
+    // that falls out of the range of 2xx
+    return `remote webhook failure`;
+  } else if (error.request) {
+    // The request was made but no response was received
+    return `unreachable remote webhook`;
+  }
+  return `unknown webhook event`;
+}
+
+// action executor
+async function executor(execOptions: ActionTypeExecutorOptions): Promise<ActionTypeExecutorResult> {
+  const log = (msg: string) => execOptions.services.log(['warn', 'actions', 'webhook'], msg);
+
+  const id = execOptions.id;
+  const { method, url, headers = {} } = execOptions.config as ActionTypeConfigType;
+  const { user: username, password } = execOptions.secrets as ActionTypeSecretsType;
+  const { body: data } = execOptions.params as ActionParamsType;
+
+  const result: Result<AxiosResponse, AxiosError> = await promiseResult(
+    axios.request({
+      method,
+      url: asComposedUrl(url),
+      auth: {
+        username,
+        password,
+      },
+      headers,
+      data,
+    })
+  );
+
+  if (isOk(result)) {
+    const { value: response } = result;
+    log(`response from ${id} webhook event: ${response.status}`);
+
+    if (response.status >= 200 && response.status < 300) {
+      return {
+        status: 'ok',
+        data: response.data,
+      };
+    } else {
+      const message = i18n.translate('xpack.actions.builtin.webhook.invalidResponseErrorMessage', {
+        defaultMessage: `an http error {code} occurred in action "{id}", due to an unknown webhook event: {data}`,
```

Comment:
These error messages can make their way back to the user.  They will likely also end up in history / audit logs.  So, we want to make them smallish, and we want to make sure we don't leak data.  For my more recent actions, I've been pretty skimpy on the error messages, not really passing anything back but http status codes, and config-schema validation messages.

Basically, we likely don't want to include `data` in this message :-)

Reply:
Fair enough, replaced with the status text.
I do see in Slack we do include the error.message in some cases though, is that not the same issue?

Reply:
Ya, I've been learning as I'm going - we might be leaking something in the older actions; feel free to fix these up if you notice them.

I added a card [standardize 3rd party http response handling](https://github.com/elastic/kibana/projects/26#card-25517891) ; I'm getting worried now that we'll need to make sure we return the http response in error results for things like 500 responses; for diagnostic purposes.

Reply:
Good call.
I tried merging a few of the things across Slack and Webhooks, but had to revert a bunch because our static i18n analysis failed miserably.

I'll have to revisit it.